File: src/main/java/com/slackworks/naether/Notation.java
package com.slackworks.naether;

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
   *
 * http://www.apache.org/licenses/LICENSE-2.0
   *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.sonatype.aether.artifact.Artifact;

import com.slackworks.naether.maven.Project;

/**
 * Helper for converting the dependency notations:
 * 
 *   groupId:artifactId:type:version
 * 
 * @author Michael Guymon
 * 
 */
public class Notation {
	
	// groupId:artifactId:extension:version
	private static Pattern notationPattern = Pattern.compile("^(.+):(.+):(.+):(.+)$");
	
	public static String generate(org.apache.maven.model.Dependency dependency) {
		StringBuffer notation = new StringBuffer()
				.append(dependency.getGroupId()).append(":")
				.append(dependency.getArtifactId()).append(":")
				.append(dependency.getType()).append(":")
				.append(dependency.getVersion());

		return notation.toString();
	}

	public static String generate( Project pom ) {
		return new StringBuilder( pom.getGroupId() )
			.append( ":" )
			.append( pom.getArtifactId() )
			.append( ":" )
			.append( pom.getType() )
			.append( ":" )
			.append( pom.getVersion() )
			.toString();
	}
	
	public static String generate(org.sonatype.aether.graph.Dependency dependency) {
		Artifact artifact = dependency.getArtifact();
		StringBuffer notation = new StringBuffer()
				.append(artifact.getGroupId()).append(":")
				.append(artifact.getArtifactId()).append(":")
				.append(artifact.getExtension()).append(":")
				.append(artifact.getVersion());

		return notation.toString();
	}
	
	public static Map<String,String> parse( String notation ) {
		Matcher matcher = notationPattern.matcher(notation);
		if ( matcher.find() ) {
			Map<String,String> notationMap = new HashMap<String,String>();
			notationMap.put( "groupId", matcher.group(1) );
			notationMap.put( "artifactId", matcher.group(2) );
			notationMap.put( "type", matcher.group(3) );
			notationMap.put( "version", matcher.group(4) );
			return notationMap;
		} else {
			return null;
		}
	}
}

File: core/src/main/ruby/naether/notation.rb
class Naether
  class Notation
    attr_reader :group, :artifact, :version, :classifier, :type
    
    PATTERN = Regexp.compile( '^(.+?):(.+?):(.+?)(:(.+?)(:(.+))?)?$' )
    
    def initialize(notation)
      if notation =~ PATTERN
        @group = Regexp.last_match(1) 
        @artifact = Regexp.last_match(2)
        
        # artifactId:groupId:type:classifier:version 
        if Regexp.last_match(7)
          @type = Regexp.last_match(3)
          @classifier = Regexp.last_match(5)
          @version = Regexp.last_match(7)
          
        # artifactId:groupId:type:version 
        elsif Regexp.last_match(5)
          @type = Regexp.last_match(3)
          @version = Regexp.last_match(5)
        # artifactId:groupId:version -
        else
          @type = 'jar'
          @version = Regexp.last_match(3)
        end
          
      end
    end
  end
end